server/repository/src/db_diesel/program_requisition/program_supplier.rs

```diff
@@ -0,0 +1,314 @@
+        master_list_row::master_list::dsl as master_list_dsl, name_row::name::dsl as name_dsl,
+        name_store_join::name_store_join::dsl as name_store_join_dsl,
+        program_row::program::dsl as program_dsl, store_row::store::dsl as store_dsl,
+    },
+    diesel_macros::apply_equal_filter,
+    repository_error::RepositoryError,
+    EqualFilter, Name, NameFilter, NameRepository, NameRow, NameStoreJoinRow, ProgramRow,
+    StorageConnection, StoreRow,
+};
+
+pub type ProgramSupplierJoin = (NameRow, NameStoreJoinRow, StoreRow, ProgramRow);
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct ProgramSupplier {
+    pub supplier: Name,
+    pub program: ProgramRow,
+}
+
+#[derive(Clone, PartialEq, Debug, Default)]
+pub struct ProgramSupplierFilter {
+    pub program_id: Option<EqualFilter<String>>,
+    pub name: Option<NameFilter>,
+}
+
+pub struct ProgramSupplierRepository<'a> {
+    connection: &'a StorageConnection,
+}
+
+impl<'a> ProgramSupplierRepository<'a> {
+    pub fn new(connection: &'a StorageConnection) -> Self {
+        ProgramSupplierRepository { connection }
+    }
+
+    pub fn query(
+        &self,
+        store_id: &str,
+        ProgramSupplierFilter {
+            program_id: program_id_filter,
+            name: name_filter,
+        }: ProgramSupplierFilter,
+    ) -> Result<Vec<ProgramSupplier>, RepositoryError> {
+        let mut query = NameRepository::create_filtered_query(store_id.to_string(), name_filter)
+            .inner_join(
+                master_list_name_join_dsl::master_list_name_join
+                    .on(master_list_name_join_dsl::name_id.eq(name_dsl::id)),
+            )
+            .inner_join(
+                master_list_dsl::master_list
+                    .on(master_list_dsl::id.eq(master_list_name_join_dsl::master_list_id)),
+            )
+            .inner_join(
+                program_dsl::program.on(program_dsl::master_list_id.eq(master_list_dsl::id)),
+            );
+
+        apply_equal_filter!(query, program_id_filter, program_dsl::id);
+
+        //  Debug diesel query
+        // println!(
+        //     "{}",
+        //     diesel::debug_query::<crate::DBType, _>(&query).to_string()
+        // );
+
+        let query = query.select((
+            // Same as NameRepository
+            name_dsl::name::all_columns(),
+            name_store_join_dsl::name_store_join::all_columns(),
+            store_dsl::store::all_columns(),
+            program_dsl::program::all_columns(),
+        ));
+        let result = query.load::<ProgramSupplierJoin>(&self.connection.connection)?;
+
+        Ok(result
+            .into_iter()
+            .map(
+                |(name_row, name_store_join_row, store_row, program)| ProgramSupplier {
+                    supplier: Name::from_join((
+                        name_row,
+                        Some(name_store_join_row),
+                        Some(store_row),
```

Comment:
although obviously rust doesn't think so...

Reply:
I had the same precaution, but was quite focused on getting this going (so forgot about it, thanks for raising), now if I change is_visible(false) in tests, I get this error:

```
Err(DBError { msg: "DIESEL_DESERIALIZATION_ERROR", extra: "UnexpectedNullError" }),
```

For now I've just made sure it would never be the case with: bd2278fcbd5ea27dda6a709d325be0db46d3bfe8

And added a self note to raise this with diesel (to test first with their latest version)